### DISCUSSION TITLE
Support for RTC

### DISCUSSION THREAD

**Discussion by dirkbr** [modm-io/modm#730] [Ideas]
It would be nice to have support for the internal RTC of the STM32F103 series.

**Comment by salkinium** [comment]
What features would the implementation require? What is your use-case for using the RTC?

**Comment by salkinium** [comment]
Ideally the API would implement the libc and libc++ interfaces for time and date, right? So the interface is kinda already defined, which is nice.

**Comment by dirkbr** [comment]
A libc and libc++ API would be perfect! One use-case is a mobile data logger, that doesn't have a permanent power supply.

**Comment by salkinium** [comment]
Note that modm is not designed for doing anything fancy with power (neither sleeping,  nor purely interrupt based operations, nor non-volatile storage, nor partial power operations), so check if you can meet your power budget *before* relying on modm too much.

**Comment by StevenMacias** [comment]
I am interested on contributing to this area. My use case is to set up DateTime in the STM32 using a communication protocol like RS422 and then being able to retrieve a timestamp with millisecond resolution. This would be handy when storing data logs in flash. 

I was thinking on using two separate variables for this purpose: 

- A 32 bit variable for storing a Unix timestamp
- A 32 bit variable for storing milliseconds calculated from the SubSeconds register. 

Similar to the `timeval` in here: https://www.daemon-systems.org/man/timespec.3.html

I am currently working on the G4 family but probably the driver can be used for other families too. How and where should I start this journey? I have contributed before with examples and fixes but this would be my first time with an actual driver contribution at modm. Any advice and recommendation is more than welcome. Thanks!

**Reply by salkinium** [reply]
You would create a new peripheral in `src/modm/platform/rtc/stm32` with a `module.lb` file that only enables the G4 family for now (we can always extend it later), then add you C++ implementation there. 
I would not worry too much about the interface, you can make it STM32 specific, the use-cases that consume the RTC data are probably so diverse that they would need a translation layer anyways.

**Reply by salkinium** [reply]
You'll probably also need to look into the RCC implementation since the low-speed clock inputs are not regularly used, so there may be some code missing to connect it to the RTC peripheral.

**Reply by StevenMacias** [reply]
Thank you so much for the insights! I will start digging into it soon.
